src/main/java/deterministicparrot/TaskList.java

```diff
@@ -0,0 +1,196 @@
+package deterministicparrot;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Represents a list of tasks that can be managed and manipulated.
+ */
+class TaskList {
+    public class SearchResult{
+        public int index;
+        public Task task;
+    }
+    private List<Task> list;
+
+    /**
+     * Constructs an empty TaskList.
+     */
+    public TaskList() {
+        this.list = new ArrayList<>();
+    }
+
+    /**
+     * Constructs a TaskList with the given list of tasks.
+     *
+     * @param tasks The list of tasks to initialize the TaskList.
+     */
+    public TaskList(List<Task> tasks) {
+        this.list = tasks;
+    }
+
+    /**
+     * Formats the tasks in the TaskList as a string.
+     *
+     * @return A formatted string representing the tasks in the TaskList.
+     */
+    public String formatAsString() {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < this.list.size(); i++) {
+            sb.append("     " + (i + 1) + ". " + this.list.get(i) + "\n");
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Serializes the TaskList to a string format.
+     *
+     * @return A serialized string representation of the TaskList.
+     */
+    public String serialize() {
+        StringBuilder sb = new StringBuilder();
+        for (Task task : this.list) {
+            if (task instanceof ToDo) {
+                sb.append("T | ").append(task.getIsDone() ? "1" : "0").append(" | ").append(task.getName()).append("\n");
+            } else if (task instanceof Deadline) {
+                Deadline deadline = (Deadline) task;
+                sb.append("D | ").append(task.getIsDone() ? "1" : "0").append(" | ").append(task.getName()).append(" | ").append(DPUtils.saveFormatDateTime(deadline.by)).append("\n");
+            } else if (task instanceof Event) {
+                Event event = (Event) task;
+                sb.append("E | ").append(task.getIsDone() ? "1" : "0").append(" | ").append(task.getName()).append(" | ").append(DPUtils.saveFormatDateTime(event.timeStart)).append(" ").append(DPUtils.saveFormatDateTime(event.timeEnd)).append("\n");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Deserializes a string representation to create a TaskList.
+     *
+     * @param rawData The serialized string representing the TaskList.
+     * @return A deserialized TaskList object.
+     * @throws Exception If there's an error during deserialization.
+     */
+    public static TaskList deserialize(String rawData) throws Exception {
```

Comment:
Good job parsing :)
Btw, if im not wrong, you dont need to indent the cases. 
E.g
```java
switch (data[0]) {
case "T":
    ToDo todo = new ToDo(data[2]);
    if (data[1].equals("1")) todo.markAsDone();
    tasks.add(todo);
    break;
case "D":
    Deadline deadline = new Deadline(data[2], data[3]);
    if (data[1].equals("1")) deadline.markAsDone();
    tasks.add(deadline);
    break;
case "E":
    String[] time = data[3].split(" ");
    Event event = new Event(data[2], time[0], time[1]);
    if (data[1].equals("1")) event.markAsDone();
    tasks.add(event);
    break;
}
```

Reply:
ah okay, thanks for that. will change later.